apps/examples/hello/hello_main.c

```diff
@@ -68,5 +70,46 @@ int hello_main(int argc, char *argv[])
 #endif
 {
 	printf("Hello, World!!\n");
+	/* Call stay IOCTL to keep PM domain in current state because maybe we know we need to do some
+	 * non-frequent work with time gaps and dont want the board to sleep in between
+	 */
+	pm_ioctl(PM_IOC_STAY);
+
+	{
+		//The work happens
+	}
+
+	/* The work is done, now relax the stay on the current state of the domain */
+	pm_ioctl(PM_IOC_RELAX);
+
+	{
+		//App does some other non-blocking work
+	}
+
+	/* Now, maybe the app has completed the major part of its work and is suggesting that the board
+	 * can go to sleep, given that there is nothing else going on.
+	 * This does not mean that the board will immediately be put to sleep. If there is a stay applied
+	 * by another application, then this SLEEP request will be DISCARDED(can be discussed if otherwise).
+	 */
+	pm_ioctl(PM_IOC_SLEEP);
+
+	/* However, as discussed before, we may need the application to request/suggest the core to go to
+	 * sleep for a specified amount of time. Alternatively, the application may want the core to be woken up at
+	 * a specific timestamp when it expects to be able to do resume some work. In this case, the application
+	 * should be able to request sleep with a timed interrupt.
+	 */
+	//pm_ioctl(PM_IOC_TIMEDSLEEP, <second argument for timed interval>, \
+			<third argument for current time so that timer interrupt can be adjusted for elapsed time>)
+
+	/* NOTE: At the beginning of this process, it seems that PM_IOC_RELAX and PM_IOC_SLEEP are similar in the
```

Comment:
Seems like you are trying to differentiate two types of low power behaviour, one with ladder-like state drop, one with immediate transition to PM_SLEEP, I think it might be feasible, since we will do most of our status check for driver in PM_SLEEP.
But, it still requires check whether the watchdog timer for PM module will be maintained well for this operation.

Reply:
I think my point is slightly different.

I mean that while PM_IOC_RELAX command will reduce the staycount and then board MAY go to sleep, PM_IOC_SLEEP WILL DIRECTLY try to invoke sleep.

However, the ladder like mechanism needs to be kept in mind. We generally do not jump to sleep state directly. But since we have nothing to do in NORMAL/IDLE, we only need to worry about the wakelock changes in STANDBY state and then we can grant the App's request for sleep.

Reply:
I was abit confused here, because from the second discussion thread counting from bottom, seems like we are expecting a non-natural way to do state transition directly to PM_SLEEP, if thats the case, how do we obey the ladder like mechanism in this situation?

Reply:
Understood, I will try to revise it accordingly.